articles/flask-crud-api/index.md

```diff
@@ -0,0 +1,306 @@
+`from flask_sqlalchemy import SQLAlchemy` imports SQLAlchemy module which we use to connect to the database.
+`from marshmallow import fields` and `from marshmallow_sqlalchemy import ModelSchema` imports the modules that we will use to serialize Python objects.
+
+### Project configuration
+In this section, we create an instance of the flask application, setup the database configurations and create a database object.
+```python
+app = Flask(__name__)
+app.config['SQLALCHEMY_DATABASE_URI'] = 'mysql+pymysql://root:@localhost:3306/todo'
+db = SQLAlchemy(app)
+```
+`app = Flask(__name__)` creates an instance of the flask web application.
+
+`app.config['SQLALCHEMY_DATABASE_URI'] = 'mysql+pymysql://<username>:<password>@localhost:
+<database port>/<database name>'
+` contains the information required to connect to our database. Our database username is `root` ,the password is ` ` ,the database port is `3306` and the database name is `todo`.
+
+On your computer start the XAMPP control panel, start the apache and MySQL servers then visit `http://localhost/phpmyadmin/index.php` in your browser and create a database with the name `todo`.
+
+`db = SQLAlchemy(app)` creates an object of the SQLAlchemy and stores it in a variable `db`.
+
+### Creating the Todo model
+We are going to the `Todo` model which will be a Python class representing the `Todo` table in the database.
+
+```python
+# Model
+class Todo(db.Model):
+    __tablename__ = "todos"
+    id = db.Column(db.Integer, primary_key=True)
+    title = db.Column(db.String(20))
+    todo_description = db.Column(db.String(100))
+
+    def create(self):
+        db.session.add(self)
+        db.session.commit()
+        return self
+
+    def __init__(self, title, todo_description):
+        self.title = title
+        self.todo_description = todo_description
+
+    def __repr__(self):
+        return f"{self.id}"
+
+
+db.create_all()
+```
+`Todo` model class has an id property which is used as the primary key.</br> 
+`db.create_all()` makes the application to create all the defined tables in the database.</br> 
+`db.session.add(self)` adds the `Todo` instance into the SQLAlchemy database connection session.</br> 
+`db.session.commit()` executes all the database operations that are available in the session.
+
+```python
+class TodoSchema(ModelSchema):
+    class Meta(ModelSchema.Meta):
+        model = Todo
+        sqla_session = db.session
+
+    id = fields.Number(dump_only=True)
+    title = fields.String(required=True)
+    todo_description = fields.String(required=True)
+```
+The `TodoSchema` class above makes it possible to return JSON from the Python objects from the SQLAlchemy.
+
+### Creating the API endpoints
+
+**GET endpoint**
+```python
+@app.route('/api/v1/todo', methods=['GET'])
+def index():
+    get_todos = Todo.query.all()
+    todo_schema = TodoSchema(many=True)
+    todos = todo_schema.dump(get_todos)
+    return make_response(jsonify({"todos": todos}))
```

Comment:
In CRUD, you have to POST data before getting it. Consider bringing the POST endpoint before GET.